### ISSUE TITLE
How to get the joints-data?

### ISSUE THREAD

**Issue by RuiqingTang** [li-ronghui/LODGE#23]
Hi,
After the inference was completed, I obtained a npy file. The motion-data stored in this file, and its shape is `(num_frames, 139)`.
It is not the joints-data. I want to know how to get it. There are usually 22 joints, so the joints-data' shape is `(num_frams, 22, 3)`. It should be noted that the order of each dimension may be different

**Comment by li-ronghui** [comment]
Hi, you can refer to the 'render.py', use 'smplx.forward' can get the joints data.

**Comment by Yjl-optimum** [comment]
> Hi, you can refer to the 'render.py', use 'smplx.forward' can get the joints data.
Hello, I referenced your render.py and got the following joint data output. Is this normal? Why is the shape obtained from SMPL not (22, 3)? Thank you.
`WARNING: You are using a SMPL model, with only 10 shape coefficients.
modata.shape is: (3072, 139)
modata.shape is: (3072, 159)
Joint_smplx data shape: (3072, 127, 3)
Joint_smpl data shape: (3072, 45, 3)
`
The code I used is as follows:
`import pickle
import numpy as np
import torch
import os
from smplx import SMPL, SMPLX
from pytorch3d.transforms import (axis_angle_to_matrix, matrix_to_axis_angle,
                                  matrix_to_quaternion, matrix_to_rotation_6d,
                                  quaternion_to_matrix, rotation_6d_to_matrix)

def ax_from_6v(q):
    assert q.shape[-1] == 6
    mat = rotation_6d_to_matrix(q)
    ax = matrix_to_axis_angle(mat)
    return ax

def motion_data_load_process(motionfile):
    if motionfile.split(".")[-1] == "pkl":
        pkl_data = pickle.load(open(motionfile, "rb"))
        smpl_poses = pkl_data["smpl_poses"]
        modata = np.concatenate((pkl_data["smpl_trans"], smpl_poses), axis=1)
        if modata.shape[1] == 69:
            hand_zeros = np.zeros([modata.shape[0], 90], dtype=np.float32)
            modata = np.concatenate((modata, hand_zeros), axis=1)
        assert modata.shape[1] == 159
        modata[:, 1] = modata[:, 1] + 0  # + 1.25
        return modata
    elif motionfile.split(".")[-1] == "npy":
        modata = np.load(motionfile)
        if len(modata.shape) == 3 and modata.shape[1] % 8 == 0:
            print("modata has 3 dim , reshape the batch to time!!!")
            modata = modata.reshape(-1, modata.shape[-1])
        if modata.shape[-1] == 315:
            print("modata.shape is:", modata.shape)
            rot6d = torch.from_numpy(modata[:, 3:])
            T, C = rot6d.shape
            rot6d = rot6d.reshape(-1, 6)
            axis = ax_from_6v(rot6d).view(T, -1).detach().cpu().numpy()
            modata = np.concatenate((modata[:, :3], axis), axis=1)
            print("modata.shape is:", modata.shape)
        elif modata.shape[-1] == 319:
            print("modata.shape is:", modata.shape)
            modata = modata[:, 4:]
            rot6d = torch.from_numpy(modata[:, 3:])
            T, C = rot6d.shape
            rot6d = rot6d.reshape(-1, 6)
            axis = ax_from_6v(rot6d).view(T, -1).detach().cpu().numpy()
            modata = np.concatenate((modata[:, :3], axis), axis=1)
        elif modata.shape[-1] == 159:
            print("modata.shape is:", modata.shape)
            print("modata.shape is:", modata.shape)
        elif modata.shape[-1] == 135:
            print("modata.shape is:", modata.shape)
            if len(modata.shape) == 3 and modata.shape[0] == 1:
                modata = modata.squeeze(0)
            rot6d = torch.from_numpy(modata[:, 3:])
            T, C = rot6d.shape
            rot6d = rot6d.reshape(-1, 6)
            axis = ax_from_6v(rot6d).view(T, -1).detach().cpu().numpy()
            hand_zeros = torch.zeros([T, 90]).to(rot6d).detach().cpu().numpy()
            modata = np.concatenate((modata[:, :3], axis, hand_zeros), axis=1)
            print("modata.shape is:", modata.shape)
        elif modata.shape[-1] == 139:
            print("modata.shape is:", modata.shape)
            modata = modata[:, 4:]
            rot6d = torch.from_numpy(modata[:, 3:])
            T, C = rot6d.shape
            rot6d = rot6d.reshape(-1, 6)
            axis = ax_from_6v(rot6d).view(T, -1).detach().cpu().numpy()
            hand_zeros = torch.zeros([T, 90]).to(rot6d).detach().cpu().numpy()
            modata = np.concatenate((modata[:, :3], axis, hand_zeros), axis=1)
            print("modata.shape is:", modata.shape)
        else:
            raise ("shape error!")

        modata[:, 1] = modata[:, 1] + 0  # + 1.25
        return modata

def extract_smpl_joints_smplx(motion_data, smpl_model):
    seq_rot = torch.tensor(motion_data, dtype=torch.float32).to('cuda')
    output = smpl_model.forward(
        betas=torch.zeros([seq_rot.shape[0], 10]).to(seq_rot.device),
        transl=seq_rot[:, :3],
        global_orient=seq_rot[:, 3:6],
        body_pose=seq_rot[:, 6:69],
        jaw_pose=torch.zeros([seq_rot.shape[0], 3]).to(seq_rot),
        leye_pose=torch.zeros([seq_rot.shape[0], 3]).to(seq_rot),
        reye_pose=torch.zeros([seq_rot.shape[0], 3]).to(seq_rot),
        left_hand_pose=torch.zeros([seq_rot.shape[0], 45]).to(seq_rot),
        right_hand_pose=torch.zeros([seq_rot.shape[0], 45]).to(seq_rot),
        expression=torch.zeros([seq_rot.shape[0], 10]).to(seq_rot),
    )

    joints = output.joints.detach().cpu().numpy()
    return joints

def extract_smpl_joints_smpl(motion_data, smpl_model):
    seq_rot = torch.tensor(motion_data, dtype=torch.float32).to('cuda')
    output = smpl_model.forward(
            betas=torch.zeros([seq_rot.shape[0], 10]).to(seq_rot.device),
            transl=seq_rot[:, :3],
            global_orient=seq_rot[:, 3:6],
            body_pose=torch.cat([seq_rot[:, 6:69], seq_rot[:, 69:72], seq_rot[:, 114:117]], dim=1)
    )
    joints = output.joints.detach().cpu().numpy()
    return joints

if __name__ == '__main__':
    SMPL_PATH = "./data/SMPL_NEUTRAL.pkl"
    SMPLX_path = "./data/SMPLX_NEUTRAL.npz"
    motion_file = os.path.join('.', 'experiments', 'Local_Module', 'FineDance_FineTuneV2_Local',
                               'samples_dod_2999_299_inpaint_soft_ddim_notranscontrol_2024-10-20-14-55-07',
                               'concat', 'npy', '211.npy')

    smpl = SMPL(SMPL_PATH).to('cuda').eval()
    smplx = SMPLX(SMPLX_path, use_pca=False, flat_hand_mean=True).eval()
    smplx.to('cuda').eval()

    motion_data = motion_data_load_process(motion_file)

    joints_smplx = extract_smpl_joints_smplx(motion_data, smplx)
    print("Joint_smplx data shape:", joints_smplx.shape)
    joints_smpl = extract_smpl_joints_smpl(motion_data, smpl)
    print("Joint_smpl data shape:", joints_smpl.shape)
`

**Comment by Yjl-optimum** [comment]
Hi, I referenced your render.py and got the following joint data output. Is this normal? Why is the shape obtained from SMPL not (22, 3)? Thank you.
`WARNING: You are using a SMPL model, with only 10 shape coefficients.
modata.shape is: (3072, 139)
modata.shape is: (3072, 159)
Joint_smplx data shape: (3072, 127, 3)
Joint_smpl data shape: (3072, 45, 3)`

`
def extract_smpl_joints_smpl(motion_data, smpl_model):
    seq_rot = torch.tensor(motion_data, dtype=torch.float32).to('cuda')
    output = smpl_model.forward(
            betas=torch.zeros([seq_rot.shape[0], 10]).to(seq_rot.device),
            transl=seq_rot[:, :3],
            global_orient=seq_rot[:, 3:6],
            body_pose=torch.cat([seq_rot[:, 6:69], seq_rot[:, 69:72], seq_rot[:, 114:117]], dim=1)
    )
    joints = output.joints.detach().cpu().numpy()
    # print("out",output)
    return joints
`
